### ISSUE TITLE
Sik-Ho Tang | Review -- BEiT V2: Masked Image Modeling with Vector-Quantized Visual Tokenizers.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#139]
Sik-Ho Tang. [Review — BEiT V2: Masked Image Modeling with Vector-Quantized Visual Tokenizers](https://towardsdev.com/review-beit-v2-masked-image-modeling-with-vector-quantized-visual-tokenizers-d7f1d9944995).

**Comment by NorbertZheng** [comment]
## Overview
BEiT V2 Improves Visual Tokenizer Using Codebook Over [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295) which uses [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374)

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/286c0966-ef56-487c-af49-65cc51976e64)
Left: [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B/16; right: [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-L/16. Top-1 fine-tuning accuracy on ImageNet (224 size).

BEiT V2: Masked Image Modeling with Vector-Quantized Visual Tokenizers
BEiT V2, by University of Chinese Academy of Sciences, and Microsoft Research,
2022 arXiv v1, Just Appeared in August 2022
Self-Supervised Learning, [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0), [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295), Image Classification

 - BEiT V2 uses visual tokenizer and codebook as model architecture.
 - Enhanced Masked Image Modeling (MIM) is used for pretraining.

**Comment by NorbertZheng** [comment]
## Vector-Quantized Knowledge Distillation (VQ-KD)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ee49b318-f188-4447-b4ca-7b4f782b441b)
Training process of visual tokenizer, which maps an image to discrete visual tokens.

### Image Representation
 - [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) is used as backbone, which splits each $224\times 224$ image into a $14\times 14$ grid of image patches, where each patch is $16\times 16$.
 - Then the image patches are flattened and linearly projected into input embeddings for [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1)s.

### Training Visual Tokenizer
 - The visual tokenizer maps an image to a sequence of discrete tokens. To be specific, the image $x$ is tokenized to $z=[z_{1},…,z_{N}] \in V$, where the vocabulary $V$ (i.e., visual codebook) contains $|V|$ discrete codes.
 - Vector-quantized knowledge distillation (VQ-KD) is proposed to train the visual tokenizer.
 - VQ-KD has two modules during training, i.e., visual tokenizer, and decoder.
 - The tokenizer is consist of a [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) encoder, and a quantizer. The tokenizer first encodes the input image to vectors. Next, the vector quantizer looks up the nearest neighbor in the codebook for each patch representation $h_{i}$.
 - Let $\\{e_{1},…,e_{|V|}\\}$ denote the codebook embeddings. For the $i$-th image patch, its quantized code is obtained by:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/67110798-4da7-4731-9b75-a94dde8dda05)
where $l2$ normalization is used for codebook lookup, which is equivalent to finding codes according to cosine similarity.
 - The $l2$-normalized codebook embeddings $l2(e_{z_{i}})$ are fed to the decoder. The decoder is also a multi-layer [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) network.
 - The output vectors $\\{o_{i}\\}$ aim at reconstructing the semantic features of a teacher model, e.g., [DINO](https://sh-tsang.medium.com/cfddbb4d3549) and CLIP. (Hope I can write a review of CLIP later.)
 - The cosine similarity between the decoder output $o_{i}$ and the teacher guidance $t_{i}$ is maximized, where $t_{i}$ denote the teacher model’s feature vector of the $i$-th image patch.
 - The training objective of VQ-KD is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/67883a85-2984-4c8e-837c-3fc1cf9aae5a)
where $sg[]$ stands for the stop-gradient operator, and $D$ represents the image data used for tokenizer training.
 - The visual tokenizer of VQ-KD is instantiated as [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B/16 for both base- and large-size pretraining.
 - In experiments, codebook size K=8192. CLIP-B/16 is used as teacher.

**Comment by NorbertZheng** [comment]
But for EEG signals, we don't have teacher model, e.g., CLIP!

**Comment by NorbertZheng** [comment]
### Codebook Collapse
 - A common issue of vector quantization training is codebook collapse. In other words, only a small proportion of codes are used.
 - The dimension of the lookup space is reduced to 32-d. The low-dimensional codebook embeddings are mapped back to higher-dimensional space before feeding into the decoder.
 - Exponential moving averages (EMAs) are used to update the codebook embeddings, which tends to be more stable in the experiments.

**Comment by NorbertZheng** [comment]
The low-dimensional codebook embeddings are mapped back to higher-dimensional space before feeding into the decoder???

**Comment by NorbertZheng** [comment]
## BEiT V2 Pretraining
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b53fd2b5-b91c-4860-b36e-aaec3026c269)
Overview of masked image modeling.

### Masked Image Modeling (MIM)
 - Masked Image Modeling (MIM) in [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295) is followed.
 - Given an input image $x$, block-wisely chosen of around 40% image patches are to be masked.
 - The masked position the is termed as $M$, and a shared learnable embedding $e[M]$ is used to replace the original image patch embeddings $e_{p_{i}}$.
 - A masked image modeling head is used to predict the visual tokens of the masked positions. A softmax classifier predicts the visual tokens:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/cb96f30e-0363-4913-932d-b73dc3cc3009)
where $x_{M}$ is the masked image, and $\\{W_{c}, b_{c}\\}$ are the classifier weights.
 - Finally, the training loss of MIM can be formulated as:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/da4dbda6-ae1d-493d-8a63-805506d8a0cb)
where $z_{i}$ means the visual tokens of the original image, and $D$ represents the pretraining images.

**Comment by NorbertZheng** [comment]
### Pretraining Global Representation
 - The CLS token is explicitly pretrained for global representation. The goal is to mitigate the discrepancy between patch-level pretraining and image-level representation aggregation.
 - As illustrated in the above figure, a **representation bottleneck** is constructed to guide the CLS token to gather information.
 - In order to pretrain the last layer’s CLS token $h_{CLS}^{L}$, it is concatenated with the intermediate $l$-th layer’s patch vectors.
 - Then, $S$ is fed to a shallow (e.g., two layers) [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) decoder and conduct masked prediction.
 - The final training loss is the summation of two terms, i.e., the original loss at the $L$-th layer, and the shallow [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) decoder’s MIM loss.

**Comment by NorbertZheng** [comment]
> Intuitively, the model favors pushing the global information to $h_{CLS}^{L}$, because the model tends to fully utilize the parameters from $l+1$-th layer to $L$-th layer, in order to decrease the additional MIM loss.

All layers need to reconstruct the masked parts, need $h_{CLS}^{L}$ to store global information. The information-flow bottleneck encourages the CLS token to obtain more reliable global representations than untrained counterparts.

 - The newly added shallow decoder is only used to pretrain the CLS token, which is discarded after pretraining.
 - The decoder network is a three-layer standard [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).

**Comment by NorbertZheng** [comment]
## Experimental Results
### Fine-Tuning for Image Classification & Semantic Segmentation
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0dd76c6d-6fda-4d29-b099-703b7079c4a4)
Fine-tuning results of image classification and semantic segmentation on ImageNet-1K and [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3).

#### Image Classification
> Base-size BEiT V2 with 300 epochs pretraining schedule reaches 85.0% top-1 accuracy, which surpasses [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295), CAE, SplitMask and PeCo by 2.1%, 1.7%, 1.4% and 0.9% respectively.

 - Moreover, BEiT V2 outperforms iBoT by 1.2%, and data2vec by 0.8%. Compared with masked distillation methods, like MVP, BEiT V2 also shows superiority.
 - Furthermore, with a longer pretraining schedule, BEiT V2 achieves 85.5% top-1 accuracy, developing a new state of the art on ImageNet-1K among self-supervised methods.

> Meanwhile, BEiT V2 using [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-L/16 with 300 epochs reaches 86.6% top-1 accuracy, which is comparable to data2vec with 1600 epochs. A longer pretraining schedule further boosts the performance to 87.3%.

#### Semantic Segmentation
 - [UPerNet](https://sh-tsang.medium.com/review-upernet-unified-perceptual-parsing-for-scene-understanding-4d03247ddad2) task layer is used and the model is fine-tuned for 160K iterations with the input resolution $512\times 512$.
 - The above table shows that BEiT V2 significantly outperforms previous self-supervised methods.
 - Moreover, using the [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-L/16 model, the performance can reach 56.7, which builds a new state-of-the-art for masked image modeling on [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3).

**Comment by NorbertZheng** [comment]
### Linear Probing
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/27f5ffeb-ef6e-46d4-aa30-1cc65a29e0ad)
Top-1 accuracy of linear probing on ImageNet-1k.

BEiT V2 surpasses [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295), CAE and MVP by 23.4%, 11.8% and 4.7% respectively.

> Additionally, BEiT V2 can also outperform [MoCo v3](https://sh-tsang.medium.com/review-moco-v3-an-empirical-study-of-training-self-supervised-vision-transformers-4a1a70c53ecd), whose pretraining gets a global representation in a contrastive learning fashion. The results indicate that BEiT V2 produces decent image-level representations.

**Comment by NorbertZheng** [comment]
### Robustness
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c9f4f0d5-3bab-4c44-8e97-2a027c298e8e)
Robustness evaluation on three ImageNet variants.

> BEiT V2 achieves dramatic gains across datasets, demonstrating the superiority of the proposed method in terms of generalization.

**Comment by NorbertZheng** [comment]
### Ablation Studies
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/49b75861-cdf6-478f-afee-4e63e30f7e27)
Ablation studies for different VQ-KD settings.

 - [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B/16 without the CLS token pretraining as the baseline model is used and pretrained for 300 epochs.
 - **Deeper decoder of VQ-KD obtains better reconstruction, but lower codebook usage and downstream task performance.**
 - Moreover, we can see that **reducing dimension for codebook lookup improves codebook utilization**.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b7615bdc-5a5f-4d8e-87f9-bd8feaf85c87)
Ablation studies for CLS token pretraining.

The shallower head (1/2-layer) performs better than the deeper head (3-layer), suggesting the shallower head (lower model capacity) pays more attention to the input CLS token than the deeper head (higher model capacity).

Moreover, the proposed method outperforms the variant without CLS token pretraining.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e035a504-00d0-4016-b673-1bd6ad13378c)
Ablation studies on VQ-KD targets.

When the teacher is [DINO](https://sh-tsang.medium.com/cfddbb4d3549), BEiT V2 respectively reaches 84.4% and 49.2% on ImageNet and [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3), respectively. The results also surpass [DINO](https://sh-tsang.medium.com/cfddbb4d3549) itself by a large margin.

When the teacher model is CLIP, BEiT V2 can get consistent improvement, demonstrating the scalability of the proposed VQ-KD.

**Comment by NorbertZheng** [comment]
BEiT V2 improves the visual tokenizer using codebook over [BEiT](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295) that uses [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374).

**Comment by NorbertZheng** [comment]
## Reference
 - [2022 arXiv v1] [BEiT V2] [BEiT V2: Masked Image Modeling with Vector-Quantized Visual Tokenizers](http://arxiv.org/abs/2208.06366).
